src/webnn/native/dml/DeviceDML.h

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 The WebNN-native Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef WEBNN_NATIVE_DEVICEDML_H_
+#define WEBNN_NATIVE_DEVICEDML_H_
+
+#include "UtilsDML.h"
+
+namespace webnn::native::dml {
+
+    inline D3D12_HEAP_PROPERTIES CreateHeapProperties(
+        D3D12_HEAP_TYPE type = D3D12_HEAP_TYPE_DEFAULT) {
+        return {type, D3D12_CPU_PAGE_PROPERTY_UNKNOWN, D3D12_MEMORY_POOL_UNKNOWN, 1, 1};
+    };
+
+    inline D3D12_RESOURCE_DESC CreateResourceDesc(
+        UINT64 width,
+        D3D12_RESOURCE_FLAGS flags = D3D12_RESOURCE_FLAG_NONE) {
+        return {D3D12_RESOURCE_DIMENSION_BUFFER, 0,    width, 1, 1, 1, DXGI_FORMAT_UNKNOWN, {1, 0},
+                D3D12_TEXTURE_LAYOUT_ROW_MAJOR,  flags};
+    };
+
+    // An adapter called the "Microsoft Basic Render Driver" is always present. This adapter is a
+    // render-only device that has no display outputs.
+    bool IsWarpAdapter(IDXGIAdapter1* pAdapter);
+
+    uint64_t RoundUpToMultiple(uint64_t value, uint64_t multiple);
+
+    class Device {
+      public:
+        HRESULT Init(DXGI_GPU_PREFERENCE gpuPreference, bool useGpu = true);
+        HRESULT InitializeGraph(ComPtr<IDMLCompiledOperator>& compiledOperator,
+                                const std::vector<std::shared_ptr<InputNode>>& inputNodes,
+                                const std::vector<Node>& outputNodes);
+        HRESULT ExecuteGraph(ComPtr<IDMLCompiledOperator>& compiledOperator,
+                             const std::vector<std::shared_ptr<InputNode>>& inputNodes,
+                             const std::vector<Node>& outputNodes,
+                             std::unordered_map<std::string, Input> namedInputs,
+                             std::unordered_map<std::string, Resource> namedOutputs);
+
+        ComPtr<IDMLDevice> GetIDMLDevice() const {
+            return mDevice;
+        }
+
+      private:
+        HRESULT CreateResourcesForCompiledOperatorInitializer();
+        HRESULT CreateResourcesForCompiledOperator();
+        void BindTemporaryResource(bool bindForInitializer = true);
+        void BindPersistentResource(bool bindForInitializer = true);
+        void CopyBufferRegion(ComPtr<ID3D12Resource> srcResource,
```

Comment:
Consider creating a higher-level `webnn::native::DMLBuffer` which encapsulates both the ID3D12Resource and the state it is in. This avoids having to pass around resource-specific stuff (or handle it) it per DML  or D3D12 operation on it.

Reply:
Do you mean something like: 
```
struct  DMLBuffer{
ComPtr<ID3D12Resource> mResource;
uint64_t size = 0;
D3D12_HEAP_TYPE type = D3D12_HEAP_TYPE_DEFAULT;
bool initialized=false;
...
}
```